<a href="https://colab.research.google.com/github/Zoe-yz/-Datawhale-Pandas/blob/main/%E3%80%90Datawhal%E3%80%91pandas_task01.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Python基础

# 1. 列表推导式与条件赋值

In [1]:
L = []

In [2]:
def my_func(x):
    return 2*x

In [3]:
for i in range(5):
    L.append(my_func(i))

In [4]:
L

[0, 2, 4, 6, 8]

事实上可以利用列表推导式进行写法上的简化： [* for i in *] 。其中，第一个 * 为映射函数，其输入为后面 i 指代的内容，第二个 * 表示迭代的对象。

In [5]:
[my_func(i) for i in range(5)]

[0, 2, 4, 6, 8]

In [6]:
[m+'_'+n for m in ['a', 'b'] for n in ['c', 'd']]

['a_c', 'a_d', 'b_c', 'b_d']

另一个实用的语法糖是带有 if 选择的条件赋值，其形式为 value = a if condition else b ：

In [7]:
value = 'cat' if 2>1 else 'dog'

In [8]:
value

'cat'

In [9]:
a, b = 'cat', 'dog'
condition = 2>1
if condition:
    value = a
else:
    value = b

In [10]:
value

'cat'

In [11]:
L = [1, 2, 3, 4, 5, 6, 7]

In [12]:
L = [x if x <=5 else 5 for x in L]

In [13]:
L

[1, 2, 3, 4, 5, 5, 5]

# 2. 匿名函数与map方法

匿名函数

In [14]:
my_func = lambda x: 2*x

In [15]:
my_func(3)

6

In [16]:
multi_para_func = lambda a, b: a + b

In [17]:
multi_para_func(1, 2)

3

In [18]:
multi_para_func = lambda a, b: a - b

In [19]:
multi_para_func(1, 2)

-1

In [20]:
multi_para_func(2, 1)

1

In [21]:
[(lambda x: 2*x)(i) for i in range(5)]

[0, 2, 4, 6, 8]

map函数，map方法

对于上述的这种列表推导式的匿名函数映射， Python 中提供了 map 函数来完成，它返回的是一个 map 对象，需要通过 list 转为列表：

In [22]:
map(lambda x: 2*x, range(5))

In [23]:
list(map(lambda x: 2*x, range(5)))

[0, 2, 4, 6, 8]

对于多个输入值的函数映射，可以通过追加迭代对象实现：

In [24]:
list(map(lambda x, y: str(x)+'_'+y, range(5), list('abcde')))

['0_a', '1_b', '2_c', '3_d', '4_e']

# 3. zip对象与enumerate方法

zip函数能够把多个可迭代对象打包成一个元组构成的可迭代对象，它返回了一个 zip 对象，通过 tuple, list 可以得到相应的打包结果：

In [25]:
L1, L2, L3 = list('abc'), list('def'), list('hij')

In [26]:
zip(L1, L2, L3)

In [27]:
list(zip(L1, L2, L3))

[('a', 'd', 'h'), ('b', 'e', 'i'), ('c', 'f', 'j')]

In [28]:
tuple(zip(L1, L2, L3))

(('a', 'd', 'h'), ('b', 'e', 'i'), ('c', 'f', 'j'))

往往会在循环迭代的时候使用到 zip 函数：

In [29]:
for i, j, k in zip(L1, L2, L3):
    print(i, j, k)

a d h
b e i
c f j


enumerate 是一种特殊的打包，它可以在迭代时绑定迭代元素的遍历序号：

In [30]:
L = list('abcd')

In [31]:
for index, value in enumerate(L):
    print(index, value)

0 a
1 b
2 c
3 d


In [32]:
for index, value in zip(range(len(L)), L):
    print(index, value)

0 a
1 b
2 c
3 d


当需要对两个列表建立字典映射时，可以利用 zip 对象：

In [33]:
dict(zip(L1, L2))

{'a': 'd', 'b': 'e', 'c': 'f'}

既然有了压缩函数，那么 Python 也提供了 * 操作符和 zip 联合使用来进行解压操作：

In [34]:
zipped = list(zip(L1, L2, L3))

In [35]:
zipped

[('a', 'd', 'h'), ('b', 'e', 'i'), ('c', 'f', 'j')]

In [36]:
list(zip(*zipped))

[('a', 'b', 'c'), ('d', 'e', 'f'), ('h', 'i', 'j')]

# Numpy基础

# np数组的构造

最一般的方法是通过 array 来构造：

In [37]:
import numpy as np

np.array([1,2,3])

array([1, 2, 3])

In [38]:
np.array(range(3))

array([0, 1, 2])

In [39]:
np.array([[1,2,3],[2,3,4]])

array([[1, 2, 3],
       [2, 3, 4]])

下面讨论一些特殊数组的生成方式：

【a】等差序列： np.linspace, np.arange

In [40]:
np.linspace(2, 8, 10)  # 起始、终止（包含）、样本个数

array([2.        , 2.66666667, 3.33333333, 4.        , 4.66666667,
       5.33333333, 6.        , 6.66666667, 7.33333333, 8.        ])

In [41]:
np.arange(2, 8.1, 0.6) # 起始、终止（不包含）、步长

array([2. , 2.6, 3.2, 3.8, 4.4, 5. , 5.6, 6.2, 6.8, 7.4, 8. ])

In [42]:
np.linspace(2, 8, 4)

array([2., 4., 6., 8.])

In [43]:
np.arange(2, 9, 2)

array([2, 4, 6, 8])

【b】特殊矩阵： zeros, eye, full

In [44]:
np.zeros((2,3)) # 传入元组表示各维度大小

array([[0., 0., 0.],
       [0., 0., 0.]])

In [45]:
np.zeros(2, 3)

TypeError: ignored

In [46]:
np.eye(3) # 3*3的单位矩阵

array([[1., 0., 0.],
       [0., 1., 0.],
       [0., 0., 1.]])

In [47]:
np.eye(3, k=1) #偏移主对角线1个单位的为单位矩阵

array([[0., 1., 0.],
       [0., 0., 1.],
       [0., 0., 0.]])

In [48]:
np.eye(5, k=2)

array([[0., 0., 1., 0., 0.],
       [0., 0., 0., 1., 0.],
       [0., 0., 0., 0., 1.],
       [0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0.]])

In [50]:
np.eye(5, 2)

array([[1., 0.],
       [0., 1.],
       [0., 0.],
       [0., 0.],
       [0., 0.]])

In [51]:
np.eye((5, 2))

TypeError: ignored

In [52]:
np.full((2,3), 5)

array([[5, 5, 5],
       [5, 5, 5]])

In [53]:
np.full((2,3), [2,2,2])

array([[2, 2, 2],
       [2, 2, 2]])

In [54]:
np.full((2,3), [2,2,2], [3,3,3])

TypeError: ignored

In [55]:
np.full((2,3), [[2,2,2], [3,3,3]])

array([[2, 2, 2],
       [3, 3, 3]])

In [56]:
np.full((2,3), [1,2])

ValueError: ignored

【c】随机矩阵： np.random

最常用的随机生成函数为 rand, randn, randint, choice ，它们分别表示0-1均匀分布的随机数组、标准正态的随机数组、随机整数组和随机列表抽样：

In [58]:
np.random.rand(4) #生成服从0-1均匀分布的三个随机数

array([0.97603324, 0.77105812, 0.61874449, 0.41317609])

In [59]:
np.random.rand(5,3) #array的每个维度大小

array([[0.00339871, 0.0253952 , 0.28052608],
       [0.90755495, 0.18704743, 0.02958394],
       [0.21079541, 0.77978479, 0.07139559],
       [0.18787401, 0.50899012, 0.52970753],
       [0.51648955, 0.99476426, 0.56991249]])

In [60]:
np.random.rand(4, 2, 3)

array([[[0.73792039, 0.05257422, 0.13293683],
        [0.07709024, 0.61523575, 0.32037267]],

       [[0.33054457, 0.69342497, 0.81183221],
        [0.94100226, 0.28251911, 0.99898848]],

       [[0.83938887, 0.26449091, 0.77397812],
        [0.80032844, 0.34050202, 0.36370028]],

       [[0.67974796, 0.04373713, 0.87917732],
        [0.59920656, 0.76469633, 0.85350783]]])

对于服从区间 𝑎 到 𝑏 上的均匀分布可以如下生成：

In [61]:
a, b = 2, 8

(b - a) * np.random.rand(5) + a

array([6.4957003 , 5.51496341, 2.54992812, 3.29337459, 2.06672063])

**randn** 生成了 𝑁(0,𝐈) 的标准正态分布

In [62]:
np.random.randn(2)

array([ 0.34903438, -0.44212789])

In [63]:
np.random.randn(2, 5)

array([[ 1.5112729 ,  0.32507158,  0.63405931,  0.99292003, -0.89096112],
       [-0.19014822,  0.98208971, -0.82450125,  1.00905426, -1.03663424]])

In [64]:
np.random.randn(3, 2, 4)

array([[[-1.99496543e-01,  7.91196630e-01, -4.61218846e-01,
          7.69369892e-01],
        [ 1.63787470e-01, -8.61339497e-01, -6.63124127e-01,
         -9.79576292e-01]],

       [[-1.24071175e+00, -1.32374185e+00,  2.53529834e-01,
          7.71139977e-01],
        [ 3.01237691e-01,  1.85284310e+00, -2.95058263e-01,
         -1.03717322e+00]],

       [[ 7.03763158e-01,  1.47558401e+00, -1.49282998e-03,
          5.74284120e-01],
        [-2.04894164e+00, -1.47379095e+00, -2.16539627e-01,
          6.10060580e-01]]])

对于服从方差为 𝜎2 均值为 𝜇 的一元正态分布可以如下生成：

In [65]:
sigma, mu = 2.5, 3

mu + np.random.randn(4) * sigma  

array([4.19733941, 0.92158694, 4.92152691, 3.94625561])

randint 可以指定生成随机整数的最小值最大值（不包含）和维度大小：

In [66]:
np.random.randint(2, 9, (3,4))

array([[7, 4, 4, 4],
       [6, 8, 7, 8],
       [8, 3, 2, 3]])

choice 可以从给定的列表中，以一定概率和方式抽取结果，当不指定概率时为均匀采样，默认抽取方式为有放回抽样：

In [67]:
list = ['a', 'b', 'c', 'd']

In [68]:
np.random.choice(list, 2, replace = True)

array(['a', 'a'], dtype='<U1')

In [69]:
np.random.choice(list, 2, replace = False)

array(['d', 'c'], dtype='<U1')

In [70]:
np.random.choice(list, (2,2))

array([['c', 'b'],
       ['b', 'c']], dtype='<U1')

当返回的元素个数与原列表相同时，等价于使用 permutation 函数，即打散原列表：

In [71]:
np.random.choice(list, 4)

array(['d', 'c', 'a', 'b'], dtype='<U1')

In [72]:
np.random.choice(list, 4, replace = False)

array(['c', 'a', 'b', 'd'], dtype='<U1')

最后，需要提到的是随机种子，它能够固定随机数的输出结果：

In [73]:
np.random.seed(0)
np.random.rand()

0.5488135039273248

In [74]:
np.random.seed(1)
np.random.rand()

0.417022004702574

# 2. np数组的变形与合并

【a】转置： T

In [75]:
np.array([[1,2],[3,4],[5,6]]).T #2*3

array([[1, 3, 5],
       [2, 4, 6]])

In [76]:
np.array([[1,2],[3,4],[5,6]]) #3*2

array([[1, 2],
       [3, 4],
       [5, 6]])

【b】合并操作： r_, c_

对于二维数组而言， r_ 和 c_ 分别表示上下合并和左右合并：

In [77]:
np.r_[np.ones((2,3)), np.ones((1,3))]

array([[1., 1., 1.],
       [1., 1., 1.],
       [1., 1., 1.]])

In [78]:
np.r_[np.ones((2,3)), np.array([[1, 2, 3]])]

array([[1., 1., 1.],
       [1., 1., 1.],
       [1., 2., 3.]])

In [79]:
np.c_[np.ones((2,3)), np.ones((2, 4))]

array([[1., 1., 1., 1., 1., 1., 1.],
       [1., 1., 1., 1., 1., 1., 1.]])

In [80]:
np.c_[np.ones((2,3)), np.array([[1,2], [2,3]])]

array([[1., 1., 1., 1., 2.],
       [1., 1., 1., 2., 3.]])

一维数组和二维数组进行合并时，应当把其视作列向量，在长度匹配的情况下只能够使用左右合并的 c_ 操作：

In [81]:
try:
    np.r_[np.array([0,0]),np.zeros((2,1))]
except Exception as e:
    Err_Msg = e

In [82]:
Err_Msg

ValueError('all the input arrays must have same number of dimensions, but the array at index 0 has 1 dimension(s) and the array at index 1 has 2 dimension(s)')

In [83]:
np.r_[np.array([0,0]),np.zeros(2)]

array([0., 0., 0., 0.])

In [84]:
np.r_[np.array([[0]]),np.zeros((2,1))]

array([[0.],
       [0.],
       [0.]])

In [85]:
np.zeros(2)

array([0., 0.])

In [86]:
np.zeros((2,1))

array([[0.],
       [0.]])

In [87]:
np.array([0,0])

array([0, 0])

In [88]:
np.c_[np.array([0,0]),np.zeros((2,3))]

array([[0., 0., 0., 0.],
       [0., 0., 0., 0.]])

In [89]:
np.r_[np.array([[0,0,0]]),np.zeros((2,3))]

array([[0., 0., 0.],
       [0., 0., 0.],
       [0., 0., 0.]])

【c】维度变换： reshape

reshape 能够帮助用户把原数组按照新的维度重新排列。在使用时有两种模式，分别为 C 模式和 F 模式，分别以逐行和逐列的顺序进行填充读取。

In [90]:
np.arange(8)

array([0, 1, 2, 3, 4, 5, 6, 7])

In [91]:
target = np.arange(8).reshape(2,4)
target

array([[0, 1, 2, 3],
       [4, 5, 6, 7]])

In [92]:
np.arange(8).reshape(4,2)

array([[0, 1],
       [2, 3],
       [4, 5],
       [6, 7]])

In [93]:
target.reshape((4,2), order = 'C') # 按照行读取和填充

array([[0, 1],
       [2, 3],
       [4, 5],
       [6, 7]])

In [94]:
target.reshape((4,2), order = 'F') # 按照列读取和填充

array([[0, 2],
       [4, 6],
       [1, 3],
       [5, 7]])

特别地，由于被调用数组的大小是确定的， reshape 允许有一个维度存在空缺，此时只需填充-1即可：

In [95]:
target.reshape(4,-1)

array([[0, 1],
       [2, 3],
       [4, 5],
       [6, 7]])

下面将 n*1 大小的数组转为1维数组的操作是经常使用的：

In [96]:
target = np.ones((3,1))
target

array([[1.],
       [1.],
       [1.]])

In [97]:
a = target.reshape(-1)
a

array([1., 1., 1.])

In [98]:
a.reshape(-1)

array([1., 1., 1.])

# 3. np数组的切片与索引

数组的切片模式支持使用 slice 类型的 start:end:step 切片，还可以直接传入列表指定某个维度的索引进行切片：

In [99]:
target = np.arange(9).reshape(3,3)
target

array([[0, 1, 2],
       [3, 4, 5],
       [6, 7, 8]])

In [100]:
target[:-1, [0,2]]

array([[0, 2],
       [3, 5]])

In [101]:
target[1:, [0,1]]

array([[3, 4],
       [6, 7]])

In [102]:
target[:, [0,1]]

array([[0, 1],
       [3, 4],
       [6, 7]])

In [103]:
target[:-1, [0,1]]

array([[0, 1],
       [3, 4]])

此外，还可以利用 np.ix_ 在对应的维度上使用布尔索引，但此时不能使用 slice 切片：

In [104]:
target[np.ix_([True, False, True], [True, False, True])]

array([[0, 2],
       [6, 8]])

In [105]:
target[np.ix_([1,2], [True, False, True])]

array([[3, 5],
       [6, 8]])

当数组维度为1维时，可以直接进行布尔索引，而无需 np.ix_ ：

In [106]:
new = target.reshape(-1)
new

array([0, 1, 2, 3, 4, 5, 6, 7, 8])

In [107]:
new[new%2==0]

array([0, 2, 4, 6, 8])

In [108]:
new[new%2!=0]

array([1, 3, 5, 7])

# 4. 常用函数

【a】 where

where 是一种条件函数，可以指定满足条件与不满足条件位置对应的填充值：

In [109]:
a = np.array([-1,1,-1,0])

In [110]:
np.where(a>0, a, 5)

array([5, 1, 5, 5])

【b】 nonzero, argmax, argmin

**这三个函数返回的都是索引**， nonzero 返回非零数的索引， argmax, argmin 分别返回最大和最小数的索引：

In [111]:
a = np.array([-2,-5,0,1,3,-1])

In [112]:
np.nonzero(a)

(array([0, 1, 3, 4, 5]),)

In [113]:
a.argmax()

4

In [114]:
a.argmin()

1

【c】 any, all

any 指当序列至少 存在一个 True 或非零元素时返回 True ，否则返回 False

all 指当序列元素 全为 True 或非零元素时返回 True ，否则返回 False

In [115]:
a = np.array([0,1])

In [116]:
a.any()

True

In [117]:
a.all()

False

【d】 cumprod, cumsum, diff

cumprod, cumsum 分别表示累乘和累加函数，返回同长度的数组， diff 表示和前一个元素做差，由于第一个元素为缺失值，因此在默认参数情况下，返回长度是原数组减1

In [118]:
a = np.array([1,2,3])

In [119]:
a.cumprod()

array([1, 2, 6])

In [120]:
a.cumsum()

array([1, 3, 6])

In [121]:
np.diff(a)

array([1, 1])

【e】 统计函数

常用的统计函数包括 max, min, mean, median, std, var, sum, quantile ，其中分位数计算是全局方法，因此不能通过 array.quantile 的方法调用：

In [122]:
target = np.arange(5)
target

array([0, 1, 2, 3, 4])

In [123]:
target.max()

4

In [124]:
target.std()

1.4142135623730951

In [125]:
np.quantile(target, 0.5) # 0.5分位数

2.0

但是对于含有缺失值的数组，它们返回的结果也是缺失值，如果需要略过缺失值，必须使用 nan* 类型的函数，上述的几个统计函数都有对应的 nan* 函数。

In [126]:
target = np.array([1, 2, np.nan])
target

array([ 1.,  2., nan])

In [127]:
target.max()

nan

In [128]:
np.nanmax(target)

2.0

In [129]:
target.min()

nan

In [130]:
np.nanmin(target)

1.0

In [131]:
np.nanquantile(target, 0.5)

1.5

In [132]:
target1 = np.array([1,3,5,9])

target2 = np.array([1,5,3,-9])

In [133]:
np.cov(target1, target2)

array([[ 11.66666667, -16.66666667],
       [-16.66666667,  38.66666667]])

In [134]:
np.corrcoef(target1, target2)

array([[ 1.        , -0.78470603],
       [-0.78470603,  1.        ]])

最后，需要说明二维 Numpy 数组中统计函数的 axis 参数，它能够进行某一个维度下的统计特征计算，当 axis=0 时结果为列的统计指标，当 axis=1 时结果为行的统计指标：

In [135]:
target = np.arange(1,10).reshape(3,-1)
target

array([[1, 2, 3],
       [4, 5, 6],
       [7, 8, 9]])

In [136]:
target.sum()

45

In [137]:
target.sum(0)

array([12, 15, 18])

In [138]:
target.sum(1)

array([ 6, 15, 24])

# 5. 广播机制

广播机制用于处理两个不同维度数组之间的操作，这里只讨论不超过两维的数组广播机制。

【a】标量和数组的操作

当一个标量和数组进行运算时，标量会自动把大小扩充为数组大小，之后进行逐元素操作

In [139]:
res = 3 * np.ones((2,2)) + 1

In [140]:
res

array([[4., 4.],
       [4., 4.]])

In [141]:
res = 1 / res

In [142]:
res

array([[0.25, 0.25],
       [0.25, 0.25]])

【b】二维数组之间的操作

当两个数组维度完全一致时，使用对应元素的操作，否则会报错，除非其中的某个数组的维度是 𝑚×1 或者 1×𝑛 ，那么会扩充其具有 1 的维度为另一个数组对应维度的大小。例如， 1×2 数组和 3×2 数组做逐元素运算时会把第一个数组扩充为 3×2 ，扩充时的对应数值进行赋值。但是，需要注意的是，如果第一个数组的维度是 1×3 ，那么由于在第二维上的大小不匹配且不为 1 ，此时报错。

In [143]:
res = np.ones((3,2))
res

array([[1., 1.],
       [1., 1.],
       [1., 1.]])

In [144]:
res * np.array([[2,3]]) #扩充第一维度为3

array([[2., 3.],
       [2., 3.],
       [2., 3.]])

In [145]:
np.array([[2],[3],[4]])

array([[2],
       [3],
       [4]])

In [146]:
res * np.array([[2],[3],[4]]) #扩充第二维度为2

array([[2., 2.],
       [3., 3.],
       [4., 4.]])

In [147]:
res * np.array([[2]]) #等价于扩充两次

array([[2., 2.],
       [2., 2.],
       [2., 2.]])

【c】一维数组与二维数组的操作

当一维数组 𝐴𝑘 与二维数组 𝐵𝑚,𝑛 操作时，等价于把一维数组视作 𝐴1,𝑘 的二维数组，使用的广播法则与【b】中一致，当 𝑘!=𝑛 且 𝑘,𝑛 都不是 1 时报错。

In [148]:
np.ones(3) + np.ones((2,3))

array([[2., 2., 2.],
       [2., 2., 2.]])

In [149]:
np.ones(3) + np.ones((2,1))

array([[2., 2., 2.],
       [2., 2., 2.]])

In [150]:
np.ones(1) + np.ones((2,3))

array([[2., 2., 2.],
       [2., 2., 2.]])

# 6. 向量与矩阵的计算 

【a】向量内积： dot

In [151]:
a = np.array([1,2,3])
b = np.array([1,3,5])
a.dot(b) 

22

【b】向量范数和矩阵范数： np.linalg.norm

In [152]:
martix_target =  np.arange(4).reshape(-1,2)
martix_target

array([[0, 1],
       [2, 3]])

In [153]:
np.linalg.norm(martix_target, 'fro')

3.7416573867739413

In [154]:
np.linalg.norm(martix_target, np.inf)

5.0

【c】矩阵乘法: @

In [155]:
a = np.arange(4).reshape(-1,2)
b = np.arange(-4,0).reshape(-1,2)
a, b

(array([[0, 1],
        [2, 3]]), array([[-4, -3],
        [-2, -1]]))

In [156]:
a @ b

array([[ -2,  -1],
       [-14,  -9]])

# 练习

# EX1: 利用列表推导式写矩阵乘法

In [157]:
M1 = np.random.rand(2,3)
M2 = np.random.rand(3,4)
M1, M2

(array([[7.20324493e-01, 1.14374817e-04, 3.02332573e-01],
        [1.46755891e-01, 9.23385948e-02, 1.86260211e-01]]),
 array([[0.34556073, 0.39676747, 0.53881673, 0.41919451],
        [0.6852195 , 0.20445225, 0.87811744, 0.02738759],
        [0.67046751, 0.4173048 , 0.55868983, 0.14038694]]))

In [158]:
res = np.empty((M1.shape[0],M2.shape[1]))
res

array([[-3.5, -1.5,  0.5,  4.5],
       [ 1. ,  5. ,  3. , -9. ]])

In [159]:
res = [[sum([M1[i][k] * M2[k][j] for k in range(M1.shape[1])]) for j in range(M2.shape[1])] for i in range(M1.shape[0])]

In [160]:
res = np.array(res)
res

array([[0.45169839, 0.41198955, 0.55713346, 0.34440275],
       [0.2388667 , 0.15483408, 0.26422035, 0.0901967 ]])

In [161]:
((M1@M2 - res) < 1e-15).all()

True

# EX2: 更新矩阵

In [162]:
A = np.arange(1,10).reshape(3, -1)
A

array([[1, 2, 3],
       [4, 5, 6],
       [7, 8, 9]])

In [163]:
B = A * (1/A).sum(1).reshape(-1,1)
B

array([[1.83333333, 3.66666667, 5.5       ],
       [2.46666667, 3.08333333, 3.7       ],
       [2.65277778, 3.03174603, 3.41071429]])

# EX3: 卡方统计量

In [164]:
np.random.seed(0)
A = np.random.randint(10, 20, (8, 5))

In [165]:
B = A.sum(0).reshape(-1, 1) * A.sum(1) / A.sum()

In [167]:
((A - B)**2 / B).sum()

ValueError: ignored

In [168]:
B = A.sum(0) * A.sum(1).reshape(-1, 1) / A.sum()

In [169]:
((A - B)**2 / B).sum()

11.842696601945802

# EX4: 改进矩阵计算的性能

In [170]:
np.random.seed(0)
m, n, p = 100, 80, 50
B = np.random.randint(0, 2, (m, p))
U = np.random.randint(0, 2, (p, n))
Z = np.random.randint(0, 2, (m, n))

**原方法：**

In [171]:
def solution(B=B, U=U, Z=Z):
    L_res = []
    for i in range(m):
        for j in range(n):
            norm_value = ((B[i]-U[:,j])**2).sum()
            L_res.append(norm_value*Z[i][j])
    return sum(L_res)

solution(B, U, Z)

100566

**尝试一：**

In [172]:
(((B.sum(1).reshape(-1, 1) - U.sum(0))**2).sum() * Z).sum()

664306956

In [173]:
(B.sum(1).reshape(-1, 1) - U.sum(0))

array([[ 3,  4,  5, ...,  1,  5,  4],
       [-1,  0,  1, ..., -3,  1,  0],
       [-1,  0,  1, ..., -3,  1,  0],
       ...,
       [-4, -3, -2, ..., -6, -2, -3],
       [ 1,  2,  3, ..., -1,  3,  2],
       [-4, -3, -2, ..., -6, -2, -3]])

In [174]:
(B.sum(1).reshape(-1, 1) - U.sum(0))**2

array([[ 9, 16, 25, ...,  1, 25, 16],
       [ 1,  0,  1, ...,  9,  1,  0],
       [ 1,  0,  1, ...,  9,  1,  0],
       ...,
       [16,  9,  4, ..., 36,  4,  9],
       [ 1,  4,  9, ...,  1,  9,  4],
       [16,  9,  4, ..., 36,  4,  9]])

In [175]:
((B.sum(1).reshape(-1, 1) - U.sum(0))**2).sum()

164148

上述方法和题目的结果100566不一致，一步一步分解来看结果，问题出在((B.sum(1).reshape(-1, 1) - U.sum(0))**2).sum()最后的sum()这里，这里应该得到一个矩阵，但是却得出了一个数

**尝试二：**

‖𝐚‖22  表示向量 𝐚 的分量平方和 ∑𝑖𝑎2𝑖，根据题目给出的这个补充信息，我们可以把∑𝑖=1𝑚∑𝑗=1𝑛‖𝐵𝑖−𝑈𝑗‖22𝑍𝑖𝑗 转换一下，例如（a-b)^2 = a^2 + b^2 - 2ab

In [176]:
(((B**2).sum(1).reshape(-1,1) + (U**2).sum(0) - 2*(B @ U)) * Z).sum()

100566

这里结果就是对的了，这个方法性能优化的地方在于没有用两个for loop嵌套

# EX5：连续整数的最大长度

**尝试一：**

In [177]:
def max_length(L):
    l = [x if x == 1 else 0 for x in np.diff(L)]
    new = np.diff(np.nonzero(l))
    res = new[new == 1]
    return len(res) + 1

In [178]:
max_length([3,2,1,2,3,4,6])

3

In [179]:
max_length([3,2,1,2,3,4,6, 8, 9, 10, 11])

5

In [180]:
L = [3,2,1,2,3,4,6, 8, 9, 10, 11]
l = [x if x == 1 else 0 for x in np.diff(L)]
l

[0, 0, 1, 1, 1, 0, 0, 1, 1, 1]

In [181]:
new = np.diff(np.nonzero(l))
new

array([[1, 1, 3, 1, 1]])

In [182]:
new[new == 1]

array([1, 1, 1, 1])

In [183]:
len(new[new == 1])

4

可以看到这里出了问题，因为有三个数都等于1，但是这三个数并不是连续的1，所以算出来的length就不对了

**尝试二：**

In [184]:
L = [3, 2, 1, 2, 3, 4, 6, 8, 9, 10, 12]
lst = np.diff(L)
lst

array([-1, -1,  1,  1,  1,  2,  2,  1,  1,  2])

In [185]:
np.where(lst != 1)

(array([0, 1, 5, 6, 9]),)

In [186]:
np.diff(np.array(np.where(lst != 1)))

array([[1, 4, 1, 3]])

In [187]:
np.diff(np.array(np.where(lst != 1))).max()

4

In [188]:
def max_len(L):
    lst = np.r_[0, np.diff(L), 0]
    new = np.array(np.where(lst != 1))
    return np.diff(new).max()

test1:

In [189]:
max_len([3, 2, 1, 2, 3, 4, 6, 8, 9, 10, 12])

4

In [190]:
lst = np.r_[0, np.diff([3, 2, 1, 2, 3, 4, 6, 8, 9, 10, 12]), 0]
lst

array([ 0, -1, -1,  1,  1,  1,  2,  2,  1,  1,  2,  0])

In [191]:
new = np.array(np.where(lst != 1))
new

array([[ 0,  1,  2,  6,  7, 10, 11]])

In [192]:
np.diff(new).max()

4

test2:

In [193]:
max_length([1,2,5,6,7])

2

In [194]:
np.diff([1,2,5,6,7])

array([1, 3, 1, 1])

In [195]:
lst = np.r_[0, np.diff([1,2,5,6,7]), 0]
lst

array([0, 1, 3, 1, 1, 0])

In [196]:
new = np.array(np.where(lst != 1))
new

array([[0, 2, 5]])

In [197]:
np.diff(new).max()

3

test3:

In [198]:
max_len([3,2,1,2,3,4,6])

4

In [199]:
lst = np.r_[0, np.diff([3,2,1,2,3,4,6]), 0]
lst

array([ 0, -1, -1,  1,  1,  1,  2,  0])

In [200]:
new = np.array(np.where(lst != 1))
new

array([[0, 1, 2, 6, 7]])

In [201]:
np.diff(new).max()

4

最终连续整数的最大长度代码为：

In [202]:
def max_len(L):
    lst = np.r_[0, np.diff(L), 0]
    new = np.array(np.where(lst != 1))
    return np.diff(new).max()